### PR TITLE
Port Files navigation to vue

### PR DESCRIPTION
**Pull request by skjnldsv**
- [x] Base Navigation view
- [x] Navigation API
- [x] Legacy views API mapper
- [x] Settings to Modal
- [x] Component testing

## Fix later
- [ ] Breadcrumbs behind navigation collapse button
- [x] Add copy to clipboard for webdav link
- [ ] Fix favorite live update

## Follow-ups
- [ ] Nuke snap.js
- [ ] Release [and update webpack config](https://github.com/nextcloud/server/pull/35772#discussion_r1058390750)
- [ ] Update vue components for apps using the Settings API 
  - [ ] https://github.com/nextcloud/text/blob/78e1e6e69f0e8752440bd58e03f74b2a98dd4876/src/files.js#L53
  - [ ] https://github.com/nextcloud/recommendations/blob/a8a58005083ef0f649d48ddd584c33ba0a0f26ee/src/main.js#L68

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -113,11 +115,13 @@
   },
   "devDependencies": {
     "@babel/node": "^7.20.0",
+    "@babel/preset-typescript": "^7.18.6",
     "@cypress/browserify-preprocessor": "^3.0.2",
     "@nextcloud/babel-config": "^1.0.0",
     "@nextcloud/cypress": "^1.0.0-beta.1",
     "@nextcloud/eslint-config": "^8.0.0",
     "@nextcloud/stylelint-config": "^2.1.2",
+    "@nextcloud/webpack-vue-config": "github:nextcloud/webpack-vue-config#master",
```
Comment (skjnldsv) [inline package.json:124]: no be replaced with newly released `webpack-vue-config` when released

File: apps/files/lib/Service/UserConfig.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+use OCP\IConfig;
+use OCP\IUser;
+use OCP\IUserSession;
+
+class UserConfig {
+	const ALLOWED_CONFIGS = [
+		[
+			'key' => 'crop_image_previews',
+			'default' => true,
+			'allowed' => [true, false],
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/files/lib/Service/UserConfig.php:35]: Isn't the key self explanatory? :see_no_evil:

File: apps/files/appinfo/routes.php [hunk 3]
```diff
@@ -39,19 +39,47 @@
 
 use OCA\Files\Controller\OpenLocalEditorController;
 
+// Legacy routes above
+/** @var $this \OC\Route\Router */
+$this->create('files_ajax_download', 'apps/files/ajax/download.php')
+	->actionInclude('files/ajax/download.php');
+$this->create('files_ajax_list', 'apps/files/ajax/list.php')
```
Comment (PVince81) [inline apps/files/appinfo/routes.php:46]: I didn't know this still existed, I made a PR to kill the list one: https://github.com/nextcloud/server/pull/35980

File: apps/files/src/views/Navigation.cy.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import FolderSvg from '@mdi/svg/svg/folder.svg'
+import ShareSvg from '@mdi/svg/svg/share-variant.svg'
+
+import NavigationService from '../services/Navigation'
+import NavigationView from './Navigation.vue'
+import router from '../router/router.js'
+
+const Navigation = new NavigationService()
+
+console.log(FolderSvg)
```
Comment (PVince81) [inline apps/files/src/views/Navigation.cy.ts:11]: leftover

File: apps/files/src/views/Settings.vue [hunk 5]
```diff
@@ -20,26 +20,115 @@
...
+		</NcAppSettingsSection>
+
+		<!-- Webdav URL-->
+		<NcAppSettingsSection id="webdav" :title="t('files', 'Webdav')">
+			<NcInputField type="text" readonly="readonly" :value="webdavUrl" />
+			<em>
+				<a :href="webdavDocs" target="_blank" rel="noreferrer noopener">
+					{{ t('files', 'Use this address to access your Files via WebDAV') }} ↗
+				</a>
+			</em>
```
Comment (PVince81) [inline apps/files/src/views/Settings.vue:53]: as discussed, if quick and easy, add a clipboard button

File: package.json [hunk 6]
```diff
@@ -20,9 +20,10 @@
     "sass": "sass --load-path core/css core/css/ apps/*/css",
     "sass:watch": "sass --watch --load-path core/css core/css/ apps/*/css",
     "sass:icons": "babel-node core/src/icons.js",
-    "cypress": "npm run cypress:e2e",
+    "cypress": "npm run cypress:e2e && npm run cypress:component",
```
Comment (PVince81) [inline package.json:23]: should we run component first ? because if one of those is broken, not much point running e2e ?

**Comment by PVince81** [comment]
- [ ] BUG: I have to manually close the sidebar to see the full file view on narrow screens:

image

**Comment by PVince81** [comment]
- [ ] UX: share expand arrow is now on the right instead of originally left, and also only appear on hover.

not sure if this is part of the original component or added here, would be good to get feedback on the expected UX here @jancborchardt 

(maybe can do a review session to discuss such issues later)

image

File: apps/files/appinfo/routes.php (hunk 3 above)
Comment (skjnldsv) [reply apps/files/appinfo/routes.php:46]: Yeah, this created an issue, lol
Story of my last 3 weeks with this PR. :see_no_evil:

**Comment by PVince81** [comment]
- [ ] URL has an additional "files": https://vvortex.local/index.php/apps/files/files?dir=/test&fileid=75 instead of https://vvortex.local/index.php/apps/files/?dir=/test&fileid=75

both links seem to work though, so if this new duplication is intended then it's fine

**Comment by PVince81** [comment]
- [ ] hide additional settings when empty

image

**Comment by skjnldsv** [comment]
> Expected: when clicking on the file list the sidebar should auto-close.

This is a global decision.
Other apps do not have this behaviour. See Talk for example :)


>  UX: share expand arrow is now on the right instead of originally left, and also only appear on hover.

This has been discussed with the design team when the component have been rewritten already. This is the new and approved behaviour :)


>  URL has an additional "files": [vvortex.local/index.php/apps/files/files?dir=/test&fileid=75](https://vvortex.local/index.php/apps/files/files?dir=/test&fileid=75) instead of [vvortex.local/index.php/apps/files/?dir=/test&fileid=75](https://vvortex.local/index.php/apps/files/?dir=/test&fileid=75)
> 
> both links seem to work though, so if this new duplication is intended then it's fine

This is intended, you can have a different view. `files` is the default and can be ommited for compatibility reasons
https://github.com/nextcloud/server/blob/ad4156a4b383fe54ea5393b62c4ff8a94c9e2a7e/apps/files/src/router/router.js#L38-L47

**Review by szaimen — APPROVED** [review]
Works very well in my testing (apart from the small things that Vincent mentioned) but didnt review the code.

**Review by PVince81 — APPROVED** [review]
:+1: no critical issues

the only one that annoys me is the left sidebar behavior but needs handling separately (nextcloud-vue)

File: apps/files/src/views/Settings.vue (hunk 5 above)
Comment (PVince81) [reply apps/files/src/views/Settings.vue:53]: => seen in follow up checklist

**Comment by PVince81** [comment]
@skjnldsv please tick the checkboxes and link to tickets for the follow ups to move forward

I did some testing and apart from my above comments I didn't find anything, it works nicely.

Great job! :tada:

File: package.json (hunk 6 above)
Comment (skjnldsv) [reply package.json:23]: Fair enough :)

File: apps/files/src/views/Settings.vue (hunk 5 above)
Comment (skjnldsv) [reply apps/files/src/views/Settings.vue:53]: image

Comment (skjnldsv) [reply apps/files/src/views/Settings.vue:67]: Figured it would be quick, it was, relatively. :grin:

**Comment by skjnldsv** [comment]
Come on! This time it's the one! Final rebase (™)

**Comment by PVince81** [comment]
:tada:

**Comment by szaimen** [comment]
🎉🎉🎉🎉🎉

File: apps/files/js/app.js [hunk 7]
```diff
@@ -51,16 +51,11 @@
 		 * Initializes the files app
 		 */
 		initialize: function() {
-			this.navigation = new OCA.Files.Navigation($('#app-navigation'));
```
Comment (skjnldsv) [inline apps/files/js/app.js:54]: @come-nc here

**Pull request merged**